Fix: Handle more error cases in getAPIInnerErrors()

gu-stav opened the pull request:
### What does it do?

Updates `getAPIInnerErrors()` to handle not only errors of type `ValidationError`, but also `ApplicationError` and `YupValidationError`.

### Why is it needed?

Lifecycle hooks can throw a variety of errors: https://github.com/strapi/strapi/blob/main/packages/core/utils/lib/errors.js So far `getAPIInnerErrors()` only handled errors of type `ValidationError` but crashed on all others, because the response format is different.

This PR allows the function to handle `ApplicationError`s. Also: the tests were broken, due to a missing `.test.` file postfix.

### How to test it?

#### ValidationError

1. Clone an entity containing a slug field: http://localhost:4000/admin/content-manager/collectionType/api::restaurant.restaurant?page=1&pageSize=10&sort=name:ASC&plugins[i18n][locale]=en
2. The slug error should be displayed

#### ApplicationError

1. Update lifecycle hooks of any content-type:

```js
const { errors } = require('@strapi/utils');

module.exports = {
  beforeUpdate() {
    throw new errors.ApplicationError('Test Message');
  },
};
```

2. Update an entity of this content-type
3. Expect the error to be displayed

#### ApplicationError when deleting an entity

1. Update lifecycle hooks of any content-type:

```js
const { errors } = require('@strapi/utils');

module.exports = {
  beforeDelete() {
    throw new errors.ApplicationError('Test Message');
  },
};
```

2. Update an entity of this content-type
3. Expect the error to be displayed

### Related issue(s)/PR(s)

- Fixes https://github.com/strapi/strapi/issues/15215
- Fixes https://github.com/strapi/strapi/issues/15017
- Fixes https://github.com/strapi/strapi/issues/12607

simotae14 commented on packages/core/helper-plugin/lib/src/utils/getAPIInnerErrors/index.js:
```diff
@@ -1,16 +1,30 @@
-        field: err.path[err.path.length - 1],
-      },
+  return getErrorPayload(error).reduce((acc, err, index) => {
+    const path = err?.path ? err?.path.join('.') : `${err.name}.${index}`;
```
why are you still checking the presence of the path in err, you already did it in the condition, I think it is better to write
```
return getErrorPayload(error).reduce((acc, err, index) => {
    const path = err?.path ? err.path.join('.') : 
```
wdyt?

gu-stav replied:
I've refactored this bit and split `ValidationError` more from `ApplicationError`. I think this will help us in the future to better handle API errors in FE.

simotae14 commented on packages/core/helper-plugin/lib/src/utils/getAPIInnerErrors/index.js:
```diff
@@ -1,16 +1,30 @@
+      defaultMessage: message,
+
+      // Only errors of type YupValidationError contain a path
+      values: err?.path
```
maybe it can be useful to create a variable just for this check to avoid to do it in every place because every time it is transpiled to `err === null || err === void 0 ? void 0 : err.path`

gu-stav replied:
Refactored in fae4e10b816f5d2b95011b8348ef63306afe8fce

ronronscelestes reviewed (dismissed):
other than Simone's review, LGTM - tested each cases, well done 💃

gu-stav commented:
@ronronscelestes @simotae14 Thanks for your feedback. I've refactored this helper a bit and introduced a new one: `handleAPIError()` which should help us in the future to handle all errors returned by the API in one place. I can't remove `getAPIInnerErrors()` at the moment, but it is now a wrapper function only and uses `handleAPIError()` under the hood.

Wdyt?

gu-stav commented:
No wait: sorry for the confusion. I'll create a follow-up PR with a better API.
